packages/@aws-cdk/aws-apprunner/lib/service.ts

```diff
@@ -0,0 +1,609 @@
+    const serviceUrl = attrs.serviceUrl;
+    const serviceStatus = attrs.serviceStatus;
+
+    class Import extends cdk.Resource {
+      public readonly serviceArn = serviceArn
+      public readonly serviceName = serviceName
+      public readonly serviceUrl = serviceUrl
+      public readonly serviceStatus = serviceStatus
+    }
+
+    return new Import(scope, id);
+  }
+  private readonly props: ServiceProps;
+  private accessRole?: iam.IRole;
+  private ecrRepository?: ecr.IRepository;
+  
+  /**
+   * The ARN of the Service.
+   * @attribute
+   */
+  readonly serviceArn: string;
+  
+  /**
+   * The ID of the Service.
+   * @attribute
+   */
+  readonly serviceId: string;
+  
+  /**
+   * The URL of the Service.
+   * @attribute
+   */
+  readonly serviceUrl: string;
+  
+  /**
+   * The status of the Service.
+   * @attribute
+   */
+  readonly serviceStatus: string;
+  
+  /**
+   * The name of the service.
+   * @attribute
+   */
+  readonly serviceName: string;
+  
+  public constructor(scope: Construct, id: string, props: ServiceProps = {}) {
+    super(scope, id);
+    // either image or code is allowed. Not both.
+    if ((!props.code && !props.image) || (props.code && props.image)) {
+      throw new Error('Either image or code is required, not both.');
+    }
+    // connection is required if code is defined
+    if (props.code && !props.connection) {
+      throw new Error('connection is required for github repository source.');
+    }
+    this.props = props;
+
+    // generate an IAM role only when ImageRepositoryType is ECR and props.role is undefined
+    this.accessRole = (this.props.image && this.props.image.type == ImageRepositoryType.ECR) ?
+      this.props.accessRole ? this.props.accessRole : this.generateDefaultRole() : undefined;
+
+    const resource = new CfnService(this, 'Resource', {
+      instanceConfiguration: {
+        cpu: props.cpu?.valueOf(),
+        memory: props.memory?.valueOf(),
+        instanceRoleArn: props.instanceRole?.roleArn,
+      },
+      sourceConfiguration: {
+        authenticationConfiguration: this.renderAuthenticationConfiguration(),
+        imageRepository: this.props.image ? this.renderImageRepositoryConfiguration(this.props.image) : undefined,
+        codeRepository: this.props.code ? this.renderCodeRepositoryConfiguration(this.props.code) : undefined,
+      },
+    });
+
+    // grant required privileges for the role
+    if (this.accessRole && this.ecrRepository) {
+      this.ecrRepository.grantPull(this.accessRole);
+    }
+    this.serviceArn = resource.attrServiceArn;
+    this.serviceId = resource.attrServiceId;
+    this.serviceUrl = resource.attrServiceUrl;
+    this.serviceStatus = resource.attrStatus;
+    this.serviceName = resource.ref;
+  }
+  private renderImageRepositoryConfiguration(assets: ContainerImage): ImageRepository {
+    // grant pull privileges
+    if (assets.ecrRepository && this.accessRole) {
+      assets.ecrRepository.grantPull(this.accessRole);
```

Comment:
Is this grant performed twice? (It looks like it would be executed again at line 472.)

Reply:
> Is this grant performed twice? (It looks like it would be executed again at line 472.)

Oh yes you are right. Thanks for the heads up. Will remove the duplicate lines.